Source/WebCore/page/Navigator.cpp

```diff
@@ -450,9 +454,21 @@ int Navigator::maxTouchPoints() const
 
 NavigatorUAData& Navigator::userAgentData() const
 {
-    if (!m_navigatorUAData)
-        m_navigatorUAData = NavigatorUAData::create();
+    RefPtr frame = this->frame();
+    if (frame && frame->page()) {
+        RefPtr client = frame->loader().client();
+        if (client->hasCustomUserAgent()) {
```

Comment:
Note: we should see if there are edge cases where the client is unaware of a custom user agent (e.g., a site specific quirk)